src/test/fuzz/util.h

```diff
@@ -512,4 +512,63 @@ void ReadFromStream(FuzzedDataProvider& fuzzed_data_provider, Stream& stream) no
     }
 }
 
+class FuzzedSocket : public MockableSocket
+{
+    FuzzedDataProvider& m_fuzzed_data_provider;
+
+public:
+    explicit FuzzedSocket(FuzzedDataProvider& fuzzed_data_provider) : m_fuzzed_data_provider{fuzzed_data_provider}
+    {
+    }
+
+    ssize_t send(const char* buf, size_t len, int flags) override
+    {
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            return len;
+        }
+        return m_fuzzed_data_provider.ConsumeIntegralInRange<ssize_t>(-1, len);
+    }
+
+    ssize_t recv(char* buf, size_t len, int flags) override
+    {
+        if (buf == nullptr || len == 0 || m_fuzzed_data_provider.ConsumeBool()) {
+            return m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+        }
+        const std::vector<uint8_t> random_bytes = m_fuzzed_data_provider.ConsumeBytes<uint8_t>(len);
```

Comment:
What is the likelihood of this returning less than `len`? The `ConsumeBytes()` comment says:

> ... If fewer than
  // |num_bytes| of data remain, returns a shorter std::vector containing all
  // of the data that's left.

I think we should make it more likely that this fuzzed `recv()` returns a partial read because this is when the "interesting" things happen in the app code as it should retry.

Reply:
Good point! Pushed an updated version addressing this.